lib/matplotlib/figure.py

```diff
@@ -227,6 +228,59 @@ def get_children(self):
                 *self.legends,
                 *self.subfigs]
 
+    def get_figure(self, root=None):
+        """
+        Return the `.Figure` or `.SubFigure` instance the (Sub)Figure belongs to.
+
+        Parameters
+        ----------
+        root : bool, default=True
+            If False, return the (Sub)Figure this artist is on.  If True,
+            return the root Figure for a nested tree of SubFigures.
+
+            .. deprecated:: 3.10
+
+                From version 3.12 *root* will default to False.
+        """
+        if self._root_figure is self:
+            # Top level Figure
+            return self
+
+        if root is None:
+            # When deprecation expires, consider removing the docstring and just
+            # inheriting the one from Artist.
+            message = ('From Matplotlib 3.12 SubFigure.get_figure will by default '
+                       'return the direct parent figure, which may be a SubFigure. '
+                       'To suppress this warning, pass the root parameter.')
+            _api.warn_deprecated('3.10', message=message)
+            root = True
+
+        if root:
+            return self._root_figure
+
+        return self._parent
+
+    def set_figure(self, fig):
+        """
+        .. deprecated:: 3.10
+            Currently this method will raise an exception if *fig* is anything other
+            than the root `.Figure` this (Sub)Figure is on.  In future it will always
+            raise an exception.
+        """
+        no_switch = ("The parent and root figures of a (Sub)Figure are set at "
+                     "instantiation and cannot be changed.")
+        if fig is self._root_figure:
+            _api.deprecation("3.10",
+                             message=(f"{no_switch} From Matplotlib 3.12 this "
+                                      "operation will raise an exception."))
+            return
+
+        raise ValueError(no_switch)
```

Comment:
I'm not 100% sure this is the best solution here, but it's based on the following considerations:

1. Since the parent figure is set at instantiation and `Artist.set_figure` does not allow switching to a different figure, `(Sub)Figure.set_figure` will always either be a no-op or raise an exception.
2. For consistency with `Artist.set_figure`, the no-op should be on the parent figure.
3. For backwards compatibility and consistency with the `(Sub)Figure.figure` property, the no-op should be on the root figure.

Reply:
Not sure we can get rid of the setter without breaking OO principles (specifically Liskov's Substitution Principle)... And not sure it is worth working around that expectation break.

The parent implementation _already_ raises if a figure has already been set.

It is a reasonable point that we _already_ break the semantics of having a "set_figure" on a "Figure"... but on the other hand, all that has really changed is "a figure is set for you at instantiation", and it is actually consistent to say "set_figure on an artist that already has a figure is an error, here is a method that we just didn't override, it will error, but hey that is expected".

Reply:
> Not sure we can get rid of the setter without breaking OO principles (specifically Liskov's Substitution Principle)... And not sure it is worth working around that expectation break.

Liskov is already broken. You can’t use a Figure everywhere where you can use an Artist. If one wanted strict Liskov Substitution Principle, one would need a more refined inheritance architecture, which is not where we are. But I see this pragmatic: As long as nobody *expects* a functionality (you would not try `figure.add_artist(figure)`) or the behavior is reasonable (you can’t set another figure if one is already set), the API is good enough.